Updates to improve convergence without sacrificing computational speed

ilyamandel opened the pull request:
The computational cost has been reduced by about 25% even though the default version now aims for a fractional mass change below 0.001 per time step, a stricter threshold by a factor of 10 than previously.

Note: I'd like to do a few more tests, but am going away for the weekend, so please review without merging this in for now.

Commit:
Updates to improve convergence without sacrificing computational speed

SimonStevenson reviewed (changes requested):
This isn't looking so good at the moment. 

![remnant_mass_relation](https://github.com/user-attachments/assets/847659fa-ff66-4403-94b0-d0a53a93c845)

Here is the command I'm using to test
```
./COMPAS -n 1 --mode SSE --initial-mass r[1.0,298,0.5] --random-seed 0 --remnant-mass-prescription FRYER2012
```
[Using FRYER2012 just to avoid randomness in the initial-final mass relation]

ilyamandel commented:
Thanks, Simon -- will work on this further on Monday, away for the rest of
the weekend.

On Sat, 1 Mar 2025, 11:37 am Simon Stevenson, ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> This isn't looking so good at the moment.
>
> remnant_mass_relation.png (view on web)
> <https://github.com/user-attachments/assets/847659fa-ff66-4403-94b0-d0a53a93c845>
>
> Here is the command I'm using to test
>
> ./COMPAS -n 1 --mode SSE --initial-mass r[1.0,298,0.5] --random-seed 0 --remnant-mass-prescription FRYER2012
>
> [Using FRYER2012 just to avoid randomness in the initial-final mass
> relation]
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/TeamCOMPAS/COMPAS/pull/1344#pullrequestreview-2652034847>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD5BVNMPKRKT7IB4HYOCHD2SD6OPAVCNFSM6AAAAABYDG5UMKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNJSGAZTIOBUG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ilyamandel commented:
OK, the immediate problem is that when LBV winds turn on, since we are using time step guesses based on the mass and radius change from the previous timestep (without LBV winds, when things were changing slowly and large time steps were allowed), we immediately, in one overly large time step with extreme mass loss, deplete most or all of the star. 

Now that the problem is clear, time to think how to address it...  Thinking...

ilyamandel commented:
OK, almost there.  There were a couple of issues.  

@jeffriley : One was that the order of operations in SSE and BSE was different.  BSE computes wind mass loss first, then evolves the stars (changing their radii and luminosities).  This makes it possible to estimate beforehand how much mass the stars will lose in winds and choose the time step appropriately, to ensure that not too high a fraction of mass is lost.  SSE actually did it in the opposite order, first evolving the stars (changing radii and luminosities), then computing mass loss rates.  That makes it impossible to account in advance for whether stars will experience LBV winds and pick the timestep appropriately (if LBV winds suddenly turn on because the star crosses the Humphries-Davidson limit, we need to take much shorter time steps to adequately resolve the mass loss rate).  It also means that stars in wide binaries may evolve differently from SSE stars.  I changed SSE to follow BSE order, which means I can precompute the time step required to reasonably calculate winds.

@SimonStevenson, @ryosuke-hirai  : The second issue is that LBV winds are *really* strong.  They can reach a couple of solar masses per year in the default prescription.  This seems unphysical.  It also makes it impossible to resolve such winds since our default minimal timestep is 1 year: we always lose more than the desired fraction of mass per time step.  I've capped LBV winds to 0.1 solar masses per year in all LBV prescriptions.

Commit:
Capped total wind mass loss rate; changed SSE call order to match BSE

ilyamandel commented:
- Capped total wind mass loss rate at MAXIMUM_WIND_MASS_LOSS_RATE (set to 0.1 Msol/yr) for all prescriptions
- Changed order of calls to stellar evolution and wind mass loss in SSE to match BSE
- Updated documentation

More testing to come tomorrow (so feel free to review, but don't merge yet)

ilyamandel commented:
Not there yet...

Commit:
Clean up

ilyamandel commented:
Ready for review, but still doing some further testing

ilyamandel commented:
SSE seems to be converged.   I am comparing pre-SN total mass as a function of ZAMS mass for two runs on which one has time steps decreased by an initial factor of 10.  Other than one time point on the very edge of an extreme winds change (to be expected), the behaviour is within acceptable margins.

The runs being compared are
```
./COMPAS --mode SSE --initial-mass r[1.0,298,0.5] --random-seed 0 
./COMPAS --mode SSE --initial-mass r[1.0,298,0.5] --random-seed 0 --timestep-multiplier 0.1
```

![image](https://github.com/user-attachments/assets/ba47ad88-e185-4489-a762-878cd059b300)

ilyamandel commented:
I checked with BSE, intentionally taking a very wide binary to avoid mass transfer, so results should be essentially identical to those above:

```
./COMPAS --initial-mass-1 r[1.0,298,0.5] --initial-mass-2 1 -a 100 --random-seed 0
./COMPAS --initial-mass-1 r[1.0,298,0.5] --initial-mass-2 1 -a 100 --random-seed 0 --timestep-multiplier 0.1
```
 
 They are:
![image](https://github.com/user-attachments/assets/27ba19dc-3080-460e-b0a1-eb414385e5c3)

SimonStevenson reviewed (approved):
Thanks for the changes Ilya. I've made a remnant mass relation and plotted a few HR diagrams and everything looked sensible. Perhaps as expected, the results are considerably different to those obtained using the previous time stepping.

jeffriley reviewed (changes requested):
Placeholder - changes coming...

jeffriley committed:
wip

jmerritt1 commented:
I'm currently testing with settings relevant to compare with Figure 7, will review soon. Thanks again for the prompt fix.

jmerritt1 commented:
![image](https://github.com/user-attachments/assets/fdb4dff8-af75-4b02-9c99-f91ac7f4e3c0)

![image](https://github.com/user-attachments/assets/7675b3f6-65bb-46af-9f6f-e9beabe7372f)

![image](https://github.com/user-attachments/assets/c27aabb4-ed64-4384-a47a-496af54df93a)

In addition to significantly changing remnant masses, and in some cases remnant types, this (Times branch) has also removed the unexplained "chatter" present in the last plot (v3.07 with change-fractions 0.005)

jeffriley committed:
Minor formatting changes/tidy-up

jeffriley commented:
@ilyamandel 

Looks good (pending testing).  I just pushed some minor formatting changes.

Some questions in the code (please remove once resolved):

BaseStar::ResolveMassLoss() - see question at BaseStar.cpp line 2808
BaseStar::CalculateMassAccretedForCO() - see question in function description (BaseStar.cpp line 2936)
BaseStar::CalculateTimestep() - see question at BaseStar.cpp line 4355

And a request:

changelog and whatsnew - you also changed the mass and radial change fractions - I think you should note that.

EDIT: Actually, ignore the second question for `BaseStar::CalculateMassAccretedForCO()` - I don't think we can/should move it.  And maybe the name isn't as bad as I thought when I first wrote the question (a while ago) - we could probably leave it.

jmerritt1 reviewed (approved):
Changes look good. Replicated the initial vs. final mass plots (Figure 7). 

Edit: I also generated HR Diagrams, where despite a 10x more lenient maximum radial-change-fraction, points on HG, and as LBV, are now less sparse. Rightward incursions into the LBV regime are also significantly limited, which is a positive side effect? Besides final masses, there are no other major changes that I'm noticing here: 
![image](https://github.com/user-attachments/assets/edc59723-548c-4c81-85ca-1f9b656b93e7)

and v3.07 for comparison:
![image](https://github.com/user-attachments/assets/b58cc670-e094-4e26-be9e-9c135eaa4279)

ilyamandel commented:
Thanks for the test, @jmerritt1 !

Is it expected that PESSIMISTIC would yield larger remnant masses than MERRITT2024 for high-mass progenitors at some metallicities?  Is this due to VMS winds (Sabhahit vs Bestenlehner)?

ilyamandel commented:
@jeffriley ,
Thank you very much for your review!
A lot of the cosmetic changes you made are to code I hadn't touched in this PR, but they look OK.
Regarding your questions:
BaseStar::ResolveMassLoss() - see question at BaseStar.cpp line 2808
- I believe the order is correct in the one place where it matters, for HG stars.
BaseStar::CalculateMassAccretedForCO() - see question in function description (BaseStar.cpp line 2936)
- I think you are correct that it could be moved, but it's not really connected to this PR, so I'd prefer not to have to repeat the various tests.  I'll leave your comment to be considered later (along with some of the other *Ilya* remarks in the code that I haven't yet addressed).
BaseStar::CalculateTimestep() - see question at BaseStar.cpp line 4355
- If a timescale is negative or zero, that means something went wrong in that calculation, and we should not be using it to set the step size.  So that choice was intentional, and no changes are needed.

I'll update the changelog and whatsnew shortly.

Commit:
Responding to Jeff's questions

Commit:
Whatsnew and Changelog updates

Commit:
Fixed complaints on zero mass or radial change fractions

jeffriley reviewed (approved):
All good - thanks @ilyamandel

ilyamandel commented:
BSE convergence tests show small differences (typically at few percent level in terms of various outcomes, such as number and types of SNe, DCOs formed, etc.).  This means there is still work to be done on convergence test, so I will not close issue #24 yet.  However, even though convergence is not yet perfect, @SimonStevenson , @jeffriley and I agreed that things are now looking sufficiently good to merge this in.

jmerritt1 commented:
> Is it expected that PESSIMISTIC would yield larger remnant masses than MERRITT2024 for high-mass progenitors at some metallicities? Is this due to VMS winds (Sabhahit vs Bestenlehner)?

Yes, despite being called PESSIMISTIC, Bestenlehner gives less mass loss at high Z, and this must be responsible for the jump above 100 Msol. While we overhaul the results and discussion in the paper, we can also decide if this is, in fact the best set of prescriptions to choose. 

You mentioned adding PESSIMISTIC to COMPAS as a `mass-loss-prescription` option, we can certainly do this, if users would find it useful to override the four on-phase options.
